### PR TITLE
New module: cobra-meta

### PR DESCRIPTION
**Pull request by CarsonJM**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

**Review by maxulysse — COMMENTED** [review]
no subcommand or other usage of this tool?

File: modules/nf-core/cobrameta/environment.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+name: "cobrameta"
```
Comment (famosab) [inline modules/nf-core/cobrameta/environment.yml:1]:
```suggestion
name: cobrameta
```

File: modules/nf-core/cobrameta/main.nf [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        cobra: \$(echo \$(cobra-meta --version 2>&1) | sed 's/^.*cobra v//' ))
+    END_VERSIONS
+    """
+
+    stub:
+    def args = task.ext.args ?: ''
+    prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    mkdir ${prefix}
+    touch ${prefix}/COBRA_extended.fasta
```
Comment (famosab) [inline modules/nf-core/cobrameta/main.nf:63]: you can also create gzipped files here but it has to look like 
```suggestion
    echo "" | gzip > ${prefix}/COBRA_extended.fasta.gz
```

File: modules/nf-core/cobrameta/meta.yml [hunk 3]
```diff
@@ -0,0 +1,109 @@
+name: "cobrameta"
```
Comment (famosab) [inline modules/nf-core/cobrameta/meta.yml:1]:
```suggestion
name: cobrameta
```

File: modules/nf-core/cobrameta/meta.yml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      tool_dev_url: https://github.com/linxingchen/cobra
+      doi: 10.1038/s41564-023-01598-2
+      licence: ["MIT"]
+
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'sample1', single_end:false ]`
```
Comment (famosab) [inline modules/nf-core/cobrameta/meta.yml:25]: Is `single_end:false` correct for your data?

File: modules/nf-core/cobrameta/tests/main.nf.test [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+        then {
+            assertAll (
+                { assert process.success },
+                { assert snapshot(
+                    process.out.get(0).findAll { it != "log" }
+                    ).match()
+                },
+                { assert file(process.out.log.get(0).get(1)).exists() }
+            )
```
Comment (famosab) [inline modules/nf-core/cobrameta/tests/main.nf.test:73]: Can you also assert the version?
```suggestion
            { assert snapshot(process.out.versions).match("version") }
            )
```

File: modules/nf-core/cobrameta/tests/main.nf.test [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                )
+                input[3] = Channel.of(
+                    [
+                        [ id:'test'],
+                        [
+                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/bam/test1_contigs.sorted.bam', checkIfExists: true)
+                        ]
+                    ]
+                )
+                input[4] = 'metaspades'
```
Comment (famosab) [inline modules/nf-core/cobrameta/tests/main.nf.test:109]: I would denote the params in an extra section directly after when

```
params {
    assembler = 'metaspades'
    mink = 27
    maxk = 127
}
```

Then you can access those here like 

```suggestion
                input[4] = params.assembler
```
and so on.

Comment (famosab) [reply modules/nf-core/cobrameta/tests/main.nf.test:109]: Ah and that also for the real test

File: modules/nf-core/cobrameta/tests/main.nf.test [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    tag "modules_nfcore"
+    tag "cobrameta"
+    tag "gunzip"
+
+    setup {
+        run("GUNZIP") {
+            script "../../gunzip/main.nf"
+            process {
+                """
+                input[0] = Channel.of(
```
Comment (famosab) [inline modules/nf-core/cobrameta/tests/main.nf.test:17]: I think you do not need the Channel.of statement. You can just give the inputs in `[ ]`

```
input[0] = [ [ id:'test' ],
file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
]
```

File: modules/nf-core/cobrameta/tests/main.nf.test [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                input[5] = 27
+                input[6] = 127
+                """
+            }
+        }
+
+        then {
+            assertAll (
+                { assert process.success },
+                { assert snapshot(
```
Comment (famosab) [inline modules/nf-core/cobrameta/tests/main.nf.test:68]: Can you check some of th eoptional files here aswell? Maybe the ones you expect to be produced by your test input. You could check fasta headers.

**Review by famosab — COMMENTED** [review]
Looking good already! I added some comments :)

**Comment by CarsonJM** [comment]
@maxulysse That's correct, no subcommands or usage

@famosab Thanks for the suggestions, I'll go through and make those fixes now!

File: modules/nf-core/cobrameta/tests/main.nf.test (hunk 7 above)
Comment (maxulysse) [reply modules/nf-core/cobrameta/tests/main.nf.test:17]: It is indeed not needed, but I like seeing input as Channels made the proper way

**Comment by CarsonJM** [comment]
@famosab @maxulysse Would either of you be able to take another look at this?

**Review by maxulysse — APPROVED** [review]

**Pull request merged**